### ISSUE TITLE
Fatal error: Unexpectedly found nil while unwrapping an Optional value | iOS

### ISSUE THREAD

**Issue by Maqcel** [781flyingdutchman/background_downloader#30]
Hi I revisited my application after a while (last used version 1.6.0) and I encountered an error which I did not recognise happened ever before. Somehow it's crashing the app on both package versions.

The code used to download the content:
```dart
    await FileDownloader().download(DownloadTask(
      url: contentUrl,
      filename: fileName,
      baseDirectory: BaseDirectory.applicationDocuments,
      directory: 'courses/${course.courseId}/${step.stepId}',
    ));
```
Migrated from `await FileDownloader.download(BackgroundDownloadTask)`.
The error that appears in the console:
```bash
[Downloader] Starting task with id 158596712
[Downloader] Starting task with id 635268047
[Downloader] Fnished task with id 158596712
[Downloader] Starting task with id 1908250544
2
background_downloader/Downloader.swift:124: Fatal error: Unexpectedly found nil while unwrapping an Optional value
* thread #1, queue = 'com.apple.main-thread', stop reason = Fatal error: Unexpectedly found nil while unwrapping an Optional value
    frame #0: 0x00000001c3dad0f8 libswiftCore.dylib`_swift_runtime_on_report
libswiftCore.dylib`:
->  0x1c3dad0f8 <+0>: ret
libswiftCore.dylib`:
    0x1c3dad0fc <+0>: b      0x1c3dad0f8               ; _swift_runtime_on_report
libswiftCore.dylib`:
    0x1c3dad100 <+0>: adrp   x8, 365860
    0x1c3dad104 <+4>: ldrb   w0, [x8, #0xabc]
Target 0: (Runner) stopped.
Lost connection to device.
```
This was taken from launch on physical device, but simulator is crashing as well. I've tried this in both release and debug same results. I'm willing to continue providing more detailed information about content and app via pm, can't do it directly here from obvious reasons.

**Comment by 781flyingdutchman** [comment]
Thanks for submitting. The location where this happens is where the url you provided as a String (your contenUrl) is converted to a URL object. That conversion fails, causing nil, then causing the unwrap error. I will change the code to catch this error more gracefully, but it also is likely that the contentUrl you use is not a proper url. Can you share or verify? In particular, if the url contains special characters, is it urlEncoded?

**Comment by Maqcel** [comment]
Currently the tasks are starting and not completing I'll try with the encoding and  5.4.4

**Comment by Maqcel** [comment]
Yeah the error was displayed to the console

**Comment by 781flyingdutchman** [comment]
Yep, the URL and any query parameters must be url encoded. Thanks for pointing out the inelegant error though!

**Comment by 781flyingdutchman** [comment]
Fixed in V5.4.5
